main.go

```diff
@@ -0,0 +1,167 @@
+package main
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+const PROMPT = "> "
+const ERROR = "ERROR"
+
+func main() {
+	scanner := bufio.NewScanner(os.Stdin)
+
+	for {
+		fmt.Print(PROMPT)
+
+		scanner.Scan()
+
+		raw_text := scanner.Text()
+
+		text := removeDuplicateSpaces(raw_text)
+
+		tokens := strings.Split(text, " ")
+
+		result, err := process(tokens)
+
+		if err != nil {
+			fmt.Println(err)
+		} else {
+			fmt.Println(text, "=", result)
+		}
+	}
+}
+
+func process(tokens []string) (result string, err error) {
+	tokens, err = calculate(tokens, "*", "/")
+	if err != nil {
+		return
+	}
+
+	tokens, err = calculate(tokens, "+", "-")
+	if err != nil {
+		return
+	}
+
+	if len(tokens) == 1 {
+		result = tokens[0]
+	} else {
+		err = errors.New(ERROR)
+	}
+
+	return
+}
+
+func calculate(tmp_tokens []string, operator_1, operator_2 string) (tokens []string, err error) {
+	tokens = tmp_tokens
+
+	for {
+		index := -1
+		index, err = find_valid_operator(tokens, operator_1, operator_2)
+		if err != nil {
+			return
+		}
+
+		if index == -1 {
+			break
+		}
+
+		var result float64
+		result, err = evaluate(tokens[index-1 : index+2])
+		if err != nil {
+			return
+		}
+
+		tokens[index+1] = fmt.Sprintf("%g", result)
+		tokens = append(tokens[:index-1], tokens[index+1:]...)
+	}
+	return
+}
+
+func evaluate(formula []string) (result float64, err error) {
+
+	left_operand, err_1 := strconv.ParseFloat(formula[0], 10)
+	right_operand, err_2 := strconv.ParseFloat(formula[2], 10)
+
+	if err_1 != nil || err_2 != nil {
+		err = errors.New(ERROR)
+		return
+	}
+
+	operator := formula[1]
+
+	switch operator {
+	case "+":
+		result, _ = add(left_operand, right_operand)
+		break
+	case "-":
+		result, _ = subtract(left_operand, right_operand)
+		break
+	case "*":
+		result, _ = multiply(left_operand, right_operand)
+		break
+	case "/":
+		result, err = divide(left_operand, right_operand)
+		break
+	default:
+		err = errors.New(ERROR)
+		return
+	}
+
+	return
+}
+
+func add(left_operand, right_operand float64) (result float64, err error) {
```

Comment:
if you define same function format for every operator, we could use Strategy Pattern by defining 
```
type operator func(leftOp, rightOp float64) (float64, error)
```

Reply:
as your comment @havinhdat :
```

// https://gist.github.com/vaskoz/10168167
type Strategy func(left_operand, right_operand float64) (float64, error)
type Strategic interface {
	SetStrategy(Strategy)
	Result() (float64, error)
}
type secretStrategy struct {
	left_operand  float64
	right_operand float64
	result        float64
	err           error
	strategy      Strategy
}

func (ss *secretStrategy) SetStrategy(s Strategy) {
	ss.strategy = s
}
func (ss *secretStrategy) Result() (float64, error) {
	ss.result, ss.err = ss.strategy(ss.left_operand, ss.right_operand)
	return ss.result, ss.err
}
func New(left_operand, right_operand float64) Strategic {
	return &secretStrategy{left_operand: left_operand, right_operand: right_operand}
}

func evaluate(formula []string) (result float64, err error) {

	left_operand, err_1 := strconv.ParseFloat(formula[0], 10)
	right_operand, err_2 := strconv.ParseFloat(formula[2], 10)

	if err_1 != nil || err_2 != nil {
		err = errors.New(ERROR)
		return
	}

	strat := New(left_operand, right_operand)
	add := func(left_operand, right_operand float64) (result float64, err error) {
		result = left_operand + right_operand
		return
	}
	subtract := func(left_operand, right_operand float64) (result float64, err error) {
		result = left_operand - right_operand
		return
	}
	multiply := func(left_operand, right_operand float64) (result float64, err error) {
		result = left_operand * right_operand
		return
	}
	divide := func(left_operand, right_operand float64) (result float64, err error) {
		if right_operand == 0 {
			err = errors.New(ERROR)
		} else {
			result = left_operand / right_operand
		}

		return
	}
	operator := formula[1]

	switch operator {
	case "+":
		strat.SetStrategy(add)
		result, _ = strat.Result()
		break
	case "-":
		strat.SetStrategy(subtract)
		result, _ = strat.Result()
		break
	case "*":
		strat.SetStrategy(multiply)
		result, _ = strat.Result()
		break
	case "/":
		strat.SetStrategy(divide)
		result, _ = strat.Result()
		break
	default:
		err = errors.New(ERROR)
		return
	}

	return
}
```